File: sdks/python/taskito/resources/health.py
"""Health checker daemon thread for worker resources."""

from __future__ import annotations

import logging
import threading
import time
from typing import TYPE_CHECKING

from taskito.async_support.helpers import run_maybe_async
from taskito.resources.definition import ResourceDefinition

if TYPE_CHECKING:
    from taskito.resources.runtime import ResourceRuntime

logger = logging.getLogger("taskito.resources")


class HealthChecker:
    """Periodically checks resource health and attempts recreation on failure.

    Runs in a single daemon thread. Each resource with a non-zero
    ``health_check_interval`` is checked independently on its own schedule.
    """

    def __init__(self, runtime: ResourceRuntime) -> None:
        self._runtime = runtime
        self._stop_event = threading.Event()
        self._thread: threading.Thread | None = None

    def start(self) -> None:
        """Start the health-check daemon thread."""
        defs = self._runtime._definitions
        has_checks = any(
            d.health_check is not None and d.health_check_interval > 0 for d in defs.values()
        )
        if not has_checks:
            return

        self._stop_event.clear()
        self._thread = threading.Thread(
            target=self._run,
            daemon=True,
            name="taskito-health-checker",
        )
        self._thread.start()

    def stop(self) -> None:
        """Signal the daemon thread to stop and wait for it."""
        self._stop_event.set()
        if self._thread is not None:
            self._thread.join(timeout=5)
            self._thread = None

    def _run(self) -> None:
        """Main loop — sleep in small increments, check resources on schedule."""
        defs = self._runtime._definitions
        next_check: dict[str, float] = {}
        fail_count: dict[str, int] = {}

        for name, defn in defs.items():
            if defn.health_check is not None and defn.health_check_interval > 0:
                next_check[name] = time.monotonic() + defn.health_check_interval
                fail_count[name] = 0

        while not self._stop_event.is_set():
            now = time.monotonic()
            for name, due in list(next_check.items()):
                if now < due:
                    continue
                if name in self._runtime._unhealthy:
                    continue

                defn = defs[name]
                healthy = self._check_one(name, defn)
                if healthy:
                    fail_count[name] = 0
                else:
                    fail_count[name] += 1
                    logger.warning(
                        "Resource '%s' health check failed (%d/%d)",
                        name,
                        fail_count[name],
                        defn.max_recreation_attempts,
                    )
                    recreated = self._runtime.recreate(name)
                    if not recreated and fail_count[name] >= defn.max_recreation_attempts:
                        self._runtime.mark_unhealthy(name)

                next_check[name] = time.monotonic() + defn.health_check_interval

            self._stop_event.wait(timeout=0.5)

    def _check_one(self, name: str, defn: ResourceDefinition) -> bool:
        """Run a single health check. Returns True if healthy."""
        if defn.health_check is None:
            return True
        try:
            instance = self._runtime._instances.get(name)
            if instance is None:
                return False
            return bool(run_maybe_async(defn.health_check(instance)))
        except Exception:
            logger.exception("Health check error for resource '%s'", name)
            return False

File: sdks/node/src/resources/health.ts
import { createLogger } from "../utils";
import type { ResourceRuntime } from "./runtime";
import type { ResourceDefinition } from "./types";

const log = createLogger("resources");

/** How often the daemon wakes to see whether any resource's check is due. */
const TICK_MS = 500;

/** Fallback for {@link ResourceDefinition.maxRecreationAttempts}. */
const DEFAULT_MAX_RECREATION_ATTEMPTS = 3;

/**
 * Periodically checks worker-resource health and recreates failing instances.
 *
 * One unref'd timer wakes every {@link TICK_MS}; each resource with a positive
 * `healthCheckIntervalMs` is checked independently on its own schedule. A
 * failed (or throwing) check triggers recreation; when recreation also fails
 * and the failure budget is spent, the resource is marked permanently
 * unhealthy and never rechecked.
 */
export class HealthChecker {
  private timer: ReturnType<typeof setInterval> | undefined;
  /** The tick currently running, if any — doubles as the overlap guard. */
  private inFlight: Promise<void> | undefined;
  /** Set by {@link HealthChecker.stop} — a stopped checker never restarts. */
  private stopped = false;
  private readonly nextDue = new Map<string, number>();
  private readonly failCount = new Map<string, number>();

  constructor(private readonly runtime: ResourceRuntime) {}

  /** Start the daemon. No-op unless some resource asked for health checks. */
  start(): void {
    if (this.timer || this.stopped) {
      return;
    }
    const checked = this.runtime.healthChecked();
    if (checked.size === 0) {
      return;
    }
    const now = Date.now();
    for (const [name, def] of checked) {
      this.nextDue.set(name, now + (def.healthCheckIntervalMs ?? 0));
      this.failCount.set(name, 0);
    }
    // The tick never rejects (see below), so the timer can't leak an
    // unhandled rejection; unref keeps it from pinning the process open.
    this.timer = setInterval(() => this.onTick(checked), TICK_MS);
    this.timer.unref();
  }

  /**
   * Stop the daemon and wait for any in-flight tick to drain, so no
   * check/recreate can mutate runtime state once this resolves — safe to run
   * worker teardown right after. Terminal and idempotent: a stopped checker
   * never restarts.
   */
  async stop(): Promise<void> {
    this.stopped = true;
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = undefined;
    }
    await this.inFlight;
  }

  /** One timer wake-up: skip when the previous tick is still in flight. */
  private onTick(checked: ReadonlyMap<string, ResourceDefinition>): void {
    if (this.inFlight || this.stopped) {
      return;
    }
    this.inFlight = this.tick(checked).finally(() => {
      this.inFlight = undefined;
    });
  }

  /** Run every due check. Catches everything — timers must survive. */
  private async tick(checked: ReadonlyMap<string, ResourceDefinition>): Promise<void> {
    try {
      for (const [name, def] of checked) {
        if (this.stopped) {
          return; // shutdown began mid-tick — leave the runtime alone
        }
        if (Date.now() < (this.nextDue.get(name) ?? 0) || this.runtime.isUnhealthy(name)) {
          continue;
        }
        await this.checkAndRepair(name, def);
        this.nextDue.set(name, Date.now() + (def.healthCheckIntervalMs ?? 0));
      }
    } catch (error) {
      log.warn(() => "health-check tick failed", error);
    }
  }

  /** Check one resource; on failure recreate it, exhausting into unhealthy. */
  private async checkAndRepair(name: string, def: ResourceDefinition): Promise<void> {
    if (await this.checkOne(name, def)) {
      this.failCount.set(name, 0);
      return;
    }
    const failures = (this.failCount.get(name) ?? 0) + 1;
    this.failCount.set(name, failures);
    const maxAttempts = def.maxRecreationAttempts ?? DEFAULT_MAX_RECREATION_ATTEMPTS;
    log.warn(`resource "${name}" health check failed (${failures}/${maxAttempts})`);
    if (this.stopped) {
      return; // shutdown began while the check ran — don't recreate into teardown
    }
    if (await this.runtime.recreate(name)) {
      this.failCount.set(name, 0); // a fresh instance starts with a clean slate
    } else if (failures >= maxAttempts) {
      this.runtime.markUnhealthy(name);
    }
  }

  /** Run a single health check. A throwing check counts as unhealthy. */
  private async checkOne(name: string, def: ResourceDefinition): Promise<boolean> {
    const { healthCheck } = def;
    if (!healthCheck) {
      return true;
    }
    const built = this.runtime.builtWorkerInstance(name);
    if (!built) {
      return false; // nothing built yet — recreation will build it
    }
    try {
      return Boolean(await healthCheck(await built));
    } catch (error) {
      log.warn(() => `health check for resource "${name}" threw`, error);
      return false;
    }
  }
}
